core/src/main/java/io/questdb/griffin/engine/functions/groupby/StringAggGroupByFunction.java

```diff
@@ -34,100 +34,132 @@
+            mapValue.putLong(valueIndex, 0);
             mapValue.putBool(valueIndex + 1, true);
         }
-        mapValue.putInt(valueIndex, sinkIndex++);
-        touchedMemorySize += sink.size();
-        assertTouchedMemoryCompliance();
     }
 
     @Override
     public void computeNext(MapValue mapValue, Record record, long rowId) {
-        final DirectUtf16Sink sink = sinks.getQuick(mapValue.getInt(valueIndex));
         final CharSequence str = arg.getStrA(record);
-        if (str != null) {
-            final int hi = sink.size();
-            final boolean nullValue = mapValue.getBool(valueIndex + 1);
-            if (!nullValue) {
-                sink.putAscii(delimiter);
-            }
-            sink.put(str);
+
+        if (str == null) {
+            return;
+        }
+
+        final long ptr = mapValue.getLong(valueIndex);
+        final boolean isNull = mapValue.getBool(valueIndex + 1);
+
+        if (isNull || ptr == 0) {
+            // First non-null value for this group
+            sinkA.of(0);
+            sinkA.put(str);
+
+            totalMemoryUsed += sinkA.length() * 2;
+            assertSizeCompliance();
+
+            mapValue.putLong(valueIndex, sinkA.ptr());
             mapValue.putBool(valueIndex + 1, false);
-            touchedMemorySize += sink.size() - hi;
-            assertTouchedMemoryCompliance();
+        } else {
+            // Group already has data, append with delimiter
+            sinkA.of(ptr);
+
+            int oldLen = sinkA.length();
+
+            sinkA.putAscii(delimiter);
+            sinkA.put(str);
+
+            int newLen = sinkA.length();
+            int chars = newLen - oldLen;
+            totalMemoryUsed += chars * 2;
+            assertSizeCompliance();
+
+            mapValue.putLong(valueIndex, sinkA.ptr());
         }
     }
 
-    private void assertTouchedMemoryCompliance() {
-        if (touchedMemorySize > maxBytes) {
-            throw CairoException.nonCritical().position(functionPosition)
-                    .put("string_agg() result exceeds max size of ").put(maxBytes).put(" bytes");
+    @Override
+    public void merge(MapValue destValue, MapValue srcValue) {
+        final long srcPtr = srcValue.getLong(valueIndex);
+        final boolean isSrcNull = srcValue.getBool(valueIndex + 1);
+
+        if (isSrcNull || srcPtr == 0) {
+            // Nothing to do
+            return;
         }
+
+        final long destPtr = destValue.getLong(valueIndex);
+        final boolean isDestNull = destValue.getBool(valueIndex + 1);
+
+        if (isDestNull || destPtr == 0) {
+            // Just copy source pointer
+            destValue.putLong(valueIndex, srcPtr);
+            destValue.putBool(valueIndex + 1, false);
+            return;
+        }
+
+        // If we got here, both groups have data - need to merge them
+        sinkA.of(destPtr);
+        sinkB.of(srcPtr);
+
+        sinkA.putAscii(delimiter);
+        sinkA.put(sinkB);
```

Comment:
`assertSizeCompliance` or a similar size check should be done before merging the strings.